hail/src/main/scala/is/hail/expr/types/physical/PCanonicalBinary.scala

```diff
@@ -13,7 +13,35 @@ class PCanonicalBinary(val required: Boolean) extends PBinary {
 
   override def byteSize: Long = 8
 
+  override def copyFromType(mb: MethodBuilder, region: Code[Region], sourcePType: PType, sourceAddress: Code[Long],
+  allowDowncast: Boolean = false, forceDeep: Boolean = false): Code[Long] = {
+    if(this == sourcePType && !forceDeep) {
+      return sourceAddress
+    }
+
+    assert(this isOfType sourcePType)
+
+    val dstAddress = mb.newField[Long]
+    val length = mb.newLocal[Int]
+
+    // srcAddress must point to data by copyFromType semantics, so no runtime null-check needed
+    if(this.required > sourcePType.required) {
```

Comment:
this isn't a downcast, so no need to check here. as you point out, `copyFromType` can only be called with present values.